docs/book/v3/features/helpers/template-variable-container.md

```diff
@@ -0,0 +1,215 @@
+use Zend\Expressive\Router\Middleware\DispatchMiddleware;
+use Zend\Expressive\Router\Middleware\ImplicitHeadMiddleware;
+use Zend\Expressive\Router\Middleware\ImplicitOptionsMiddleware;
+use Zend\Expressive\Router\Middleware\MethodNotAllowedMiddleware;
+use Zend\Expressive\Router\Middleware\RouteMiddleware;
+use Zend\Stratigility\Middleware\ErrorHandler;
+
+use function Zend\Stratigility\path;
+
+return function (Application $app, MiddlewareFactory $factory, ContainerInterface $container) : void {
+    $app->pipe(ErrorHandler::class);
+    $app->pipe(ServerUrlMiddleware::class);
+
+		// The following entry is specific to this example:
+    $app->pipe(path(
+        '/api/doc',
+        $factory->lazy(TemplateVariableContainerMiddleware::class)
+    ));
+
+    $app->pipe(RouteMiddleware::class);
+
+    $app->pipe(ImplicitHeadMiddleware::class);
+    $app->pipe(ImplicitOptionsMiddleware::class);
+    $app->pipe(MethodNotAllowedMiddleware::class);
+    $app->pipe(UrlHelperMiddleware::class);
+
+    $app->pipe(DispatchMiddleware::class);
+
+    $app->pipe(NotFoundHandler::class);
+};
+```
+
+Any middleware or handler that responds to a path beginning with `/api/doc` will
+now have a `Zend\Expressive\Helper\Template\TemplateVariableContainer` attribute
+that contains an instance of that class.
+
+Within middleware that responds on that path, you can then do the following:
+
+```php
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Zend\Expressive\Helper\Template\TemplateVariableContainer;
+use Zend\Expressive\Router\RouteResult;
+
+class InjectUserAndRouteVariablesMiddleware implements MiddlewareInterface
+{
+		public function process(
+				ServerRequestInterface $request,
+				 RequestHandlerInterface $handler
+			) : ResponseInterface {
+					$container = $request->getAttribute(
+							TemplateVariableContainer::class,
+							new TemplateVariableContainer()
+					);
+
+					// Since containers are immutable, we re-populate the request:
+					$request = $request->withAttribute(
+							TemplateVariableContainer::class,
+							$container->merge([
+									'user'  => $user,
+									'route' => $request->getAttribute(RouteResult::class),
+							])
+					);
+
+					return $handler->handle($request);
+			}
+}
+```
+
+In a handler, you will call `mergeForTemplate()` with any local variables you
+want to use, including those that might override the defaults:
+
+```php
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use Zend\Expressive\Helper\Template\TemplateVariableContainer;
+use Zend\Expressive\Template\TemplateRendererInterface;
+
+class SomeHandler implements RequestHandlerInterface
+{
+		private $renderer;
```

Comment:
A bit too much indentation. 😉

Reply:
Yikes! Not sure how that happened, but I'll fix it.

Reply:
Fixed - somehow, my editor was using tabs instead of spaces.